plugins/tc/tbf/main.go

```diff
@@ -0,0 +1,208 @@
+		return nil, fmt.Errorf("if rate is set, burst must also be set")
+	case conf.Rate == 0 && conf.Burst != 0:
+		return nil, fmt.Errorf("if burst is set, rate must also be set")
+	}
+
+	return &conf, nil
+
+}
+
+func getIfbDeviceName(networkName string, containerId string) (string, error) {
+	hash := sha1.New()
+	_, err := hash.Write([]byte(networkName + containerId))
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", hash.Sum(nil))[:15], nil
+}
+
+func getMTU(deviceName string) (int, error) {
+	link, err := netlink.LinkByName(deviceName)
+	if err != nil {
+		return -1, err
+	}
+
+	return link.Attrs().MTU, nil
+}
+
+// cmdAdd is called for ADD requests
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	//no traffic shaping was requested, so just no-op and quit
+	if conf.Rate == 0 && conf.Burst == 0 {
+		return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("must be called as chained plugin")
+	}
+
+	containerIPConfigs := make([]*current.IPConfig, 0, len(conf.PrevResult.IPs))
+
+	for _, ip := range conf.PrevResult.IPs {
+		if ip.Interface == nil {
+			continue
+		}
+		intIdx := *ip.Interface
+		// Every IP is indexed in to the interfaces array, with "-1" standing
+		// for an unknown interface (which we'll assume to be Container-side
+		// Skip all IPs we know belong to an interface with the wrong name.
+		if intIdx >= 0 && intIdx < len(conf.PrevResult.Interfaces) && conf.PrevResult.Interfaces[intIdx].Name != args.IfName {
+			continue
+		}
+		containerIPConfigs = append(containerIPConfigs, ip)
+	}
+
+	if len(containerIPConfigs) == 0 {
+		return fmt.Errorf("got no container IPs")
+	}
+
+	for _, prevIface := range conf.PrevResult.Interfaces {
+		if prevIface.Sandbox != "" {
+			continue
+		}
+		err := CreateIngressQdisc(conf.Rate, conf.Burst, prevIface.Name)
+
+		if err != nil {
+			return err
+		}
+
+		mtu, err := getMTU(prevIface.Name)
+		if err != nil {
+			return err
+		}
+
+		ifbDeviceName, err := getIfbDeviceName(conf.Name, args.ContainerID)
+		if err != nil {
+			return err
+		}
+
+		err = CreateIfb(ifbDeviceName, mtu)
+		if err != nil {
+			return err
+		}
+
+		ifbDevice, err := netlink.LinkByName(ifbDeviceName)
+
+		if err != nil {
+			return err
+		}
+
+		conf.PrevResult.Interfaces = append(conf.PrevResult.Interfaces, &current.Interface{
+			Name:    ifbDeviceName,
+			Sandbox: prevIface.Sandbox,
+			Mac:     ifbDevice.Attrs().HardwareAddr.String(),
+		})
+		err = CreateEgressQdisc(conf.Rate, conf.Burst, prevIface.Name, ifbDeviceName)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Pass through the result for the next plugin
+	return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+}
+
+// cmdDel is called for DELETE requests
+func cmdDel(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+	_ = conf
```

Comment:
?

Reply:
copy-pasta from the plugin template example :-(